gobblin-yarn/src/main/java/org/apache/gobblin/yarn/YarnService.java

```diff
@@ -610,6 +610,9 @@ private boolean shouldStickToTheSameNode(int containerExitStatus) {
    */
   private void handleContainerCompletion(ContainerStatus containerStatus) {
     Map.Entry<Container, String> completedContainerEntry = this.containerMap.remove(containerStatus.getContainerId());
+    if(completedContainerEntry == null) {
+      LOGGER.warn(String.format("Container %s already been removed from containerMap before completed, there may be something wrong when starting the container", containerStatus.getContainerId()));
+    }
     String completedInstanceName = completedContainerEntry.getValue();
```

Comment:
We may still get a NPE on line 616 if completedContainerEntry is null. I think the fix probably should be to not remove the instance from the containerMap inside NMClientCallbackHandler#onStartContainerError(), since handleContainerCompletion() should be called even if there is a container launch error. Also, the double removal is preventing a replacement container being launched from handleContainerCompletion() method.